contributors/design-proposals/service_links.md

```diff
@@ -0,0 +1,318 @@
+need to consider backward compatibility when making new API changes.  As a
+requirement, backward compatibility imposes a burden that you have to live with
+the APIs you release.  This means that we should attempt to be conservative with
+new API features and avoid unnecessary complexity.  Therefore, a simpler API is
+advantageous because complexity can be added later **if desired and warranted**.
+
+#### Middleground
+
+A conservative middleground might be an API that has behavior extremely close to
+the current implicit behavior but allows specifying custom variable names.  This
+approach has the advantage of being close to what users already expect, and can
+be implemented with a fairly simple API.
+
+### Non-existent Services
+
+When a service referenced using this API does not exist, the Kubelet should
+avoid starting the pod.
+
+## Proposed Changes
+
+### API changes
+
+#### Expressing dependencies on services
+
+```go
+type PodSpec {
+  // other fields omitted
+
+  // ServiceDependencies is a list of services in this pod's namespace that the
+  // pod depends on in order to start.
+  ServiceDependencies []LocalObjectReference `json:"serviceDependencies,omitempty"`
+}
+```
+
+#### Environment variable injection
+
+Pull request [#148](https://github.com/kubernetes/community/pull/148)
+proposes an `EnvFrom` addition to the PodSpec to handle creating multiple
+environment variables from a single subject.  This API surface is ideal for the
+service use-cases, so we will extend it here:
+
+```go
+type Container struct {
+	// other fields omitted
+
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty"`
+}
+
+type EnvVarFromSource struct {
+	// other fields omitted
+
+	// Prefix is an optional prefix for generated environment variables
+	Prefix string `json:"prefix,omitempty"`
+
+	// Can be refactored to ObjectReference at a later time while
+	// maintaining backward compatibility.
+	Service *ServiceEnvSource `json:"service,omitempty"`
+}
+
+// ServiceEnvVarSource selects a Service to populate environment
+// variables with.
+//
+// The following environment variables will be generated for the
+// target service:
+//
+// <service name>_SERVICE_HOST
+// <service name>_SERVICE_PORT
+// <service name>_PORT=tcp://10.0.0.11:6379
+// <service name>_PORT_6379_TCP=tcp://10.0.0.11:6379
+// <service name>_PORT_6379_TCP_PROTO=tcp
+// <service name>_PORT_6379_TCP_PORT=6379
+// <service name>_PORT_6379_TCP_ADDR=10.0.0.11
+type ServiceEnvSource struct {
+  // The Service to generate environment variables for.
+  LocalObjectReference
+}
+```
+
+When a pod uses this API feature, it indicates that the automatic injection
+behavior is not desired for that pod.  No automatic injection will occur when
+any container in a pod uses an explicit link to a service.
+
+Notes:
+
+1. The `kubernetes` and `kubernetes-ro` services will always be injected,
+   regardless of the state of the pod's spec.
+2. The `ServiceDependencies` and `EnvFrom` API features are orthogonal
```

Comment:
If I have no ServiceDependencies and use ServiceEnvFrom, isn't that redundant since I should get all services?